5.machine_learning/3.assess_generalizability/scripts/6.compare_models_plate3.py

```diff
@@ -0,0 +1,300 @@
+plate_3_df.head()
+
+
+# In[6]:
+
+
+# Initialize empty lists to store data for each iteration
+precision_list = []
+recall_list = []
+threshold_list = []
+model_type_list = []
+feature_set_list = []
+treatment_list = []  # List for Metadata_treatment
+cell_type_list = []  # List for Metadata_cell_type
+
+# Loop through the dictionary of models (actin_final, actin_shuffled, etc.)
+for category, models in loaded_models.items():
+    # Extract model type from the category (actin_final -> final, actin_shuffled -> shuffled, etc.)
+    model_type = category.split("_")[1]
+
+    # Extract feature set from the first part of the category (actin or rest or allfeatures)
+    feature_set = category.split("_")[0]
+
+    # Iterate through each model in the current category (actin, rest, or allfeatures)
+    for model_info in models:
+        model = model_info["model"]  # Access the model object from the dictionary
+
+        print(f"Evaluating {model_type} model in {category} category...")
+
+        # Separate metadata columns
+        metadata_columns = [
+            col for col in plate_3_df.columns if col.startswith("Metadata_")
+        ]
+
+        # Separate features columns (filtering out features containing both "Correlation" and "Actin")
+        filtered_columns = [
+            col
+            for col in plate_3_df.columns
+            if not col.startswith("Metadata_")
+            and not ("Actin" in col and "Correlation" in col)
+        ]
+
+        # Select feature columns based on the feature_set type
+        if feature_set == "allfeatures":
+            # For allfeatures, use all non-metadata columns (no filtering)
+            features_columns = [
+                col for col in plate_3_df.columns if not col.startswith("Metadata_")
+            ]
+        elif feature_set == "actin":
+            features_columns = [col for col in filtered_columns if "Actin" in col]
+        elif feature_set == "rest":
+            features_columns = [
+                col for col in filtered_columns if "Actin" not in col
+            ]
+        else:
+            raise ValueError(f"Unknown feature set: {feature_set}")
+
+        # Create the new updated dataframe with the correct features
+        updated_plate_3_df = plate_3_df[metadata_columns + features_columns]
+        print(updated_plate_3_df.shape)
+
+        # Get X and y data from the dataframe
+        X, y = get_X_y_data(df=updated_plate_3_df, label=label, shuffle=False)
+
+        # Load in label encoder
+        le = load(encoder_path)
+
+        # Assign y classes to correct binary using label encoder results
+        y_binary = le.transform(y)
+
+        # Predict class probabilities for morphology feature data
+        predicted_probs = model.predict_proba(X)
+
+        # Calculate the precision, recall data
+        precision, recall, threshold = precision_recall_curve(
+            y_binary, predicted_probs[:, -1]
+        )
+        threshold = np.append(threshold, np.nan)
```

Comment:
Sorry if I'm missing something obvious, but are you also calculating AUPR somewhere?

Reply:
My bad! I did miss that, I have added it as the last code cell to the notebook, thanks!